packages/react-reconciler/src/__tests__/useMemoCache-test.js

```diff
@@ -0,0 +1,367 @@
+        return {hasError: true};
+      }
+
+      componentDidCatch(error, errorInfo) {}
+
+      render() {
+        if (this.state.hasError) {
+          // You can render any custom fallback UI
+          return <h1>Something went wrong.</h1>;
+        }
+
+        return this.props.children;
+      }
+    }
+    ErrorBoundary = _ErrorBoundary;
+  });
+
+  // @gate experimental || www
+  test('render component using cache', async () => {
+    function Component(props) {
+      const cache = useMemoCache(1);
+      expect(cache).toBeInstanceOf(Array);
+      expect(cache.length).toBe(1);
+      expect(cache[0]).toBe(undefined);
+      return 'Ok';
+    }
+    const root = ReactNoop.createRoot();
+    await act(async () => {
+      root.render(<Component />);
+    });
+    expect(root).toMatchRenderedOutput('Ok');
+  });
+
+  // @gate experimental || www
+  test('update component using cache', async () => {
+    let setX;
+    let forceUpdate;
+    function Component(props) {
+      const cache = useMemoCache(4);
+
+      // x is used to produce a `data` object passed to the child
+      const [x, _setX] = useState(0);
+      setX = _setX;
+      const c_x = x !== cache[0];
+      cache[0] = x;
+
+      // n is passed as-is to the child as a cache breaker
+      const [n, setN] = useState(0);
+      forceUpdate = () => setN(a => a + 1);
+      const c_n = n !== cache[1];
+      cache[1] = n;
+
+      let data;
+      if (c_x) {
+        data = cache[2] = {text: `Count ${x}`};
+      } else {
+        data = cache[2];
+      }
+      if (c_x || c_n) {
+        return (cache[3] = <Text data={data} n={n} />);
+      } else {
+        return cache[3];
+      }
+    }
+    let data;
+    const Text = jest.fn(function Text(props) {
+      data = props.data;
+      return data.text;
+    });
+
+    const root = ReactNoop.createRoot();
+    await act(async () => {
+      root.render(<Component />);
+    });
+    expect(root).toMatchRenderedOutput('Count 0');
+    expect(Text).toBeCalledTimes(1);
+    const data0 = data;
+
+    // Changing x should reset the data object
+    await act(async () => {
+      setX(1);
+    });
+    expect(root).toMatchRenderedOutput('Count 1');
+    expect(Text).toBeCalledTimes(2);
+    expect(data).not.toBe(data0);
+    const data1 = data;
+
+    // Forcing an unrelated update shouldn't recreate the
+    // data object.
+    await act(async () => {
+      forceUpdate();
+    });
+    expect(root).toMatchRenderedOutput('Count 1');
+    expect(Text).toBeCalledTimes(3);
+    expect(data).toBe(data1); // confirm that the cache persisted across renders
+  });
+
+  // @gate experimental || www
+  test('update component using cache with setstate during render', async () => {
+    let setX;
+    let setN;
+    function Component(props) {
+      const cache = useMemoCache(4);
+
+      // x is used to produce a `data` object passed to the child
+      const [x, _setX] = useState(0);
```

Comment:
Should we add a test for cached values that aren't primitives? Just to be sure this comparison still works:

```js
const initialTasks = [
  { id: 0, text: "Buy milk", done: true },
  { id: 1, text: "Eat tacos", done: false },
  { id: 2, text: "Brew tea", done: false },
];
function Component() {
  const cache = useMemoCache(6);
  const [tasks, setTasks] = useState(initialTasks);
  const c_tasks = cache[2] !== tasks;
  // ...
}
```